protocol-gateway/mqtt-protocol-gateway-template/src/main/java/org/eclipse/hono/gateway/sdk/mqtt2amqp/MultiTenantConnectionManagerImpl.java

```diff
@@ -0,0 +1,141 @@
+                    if (failedTenant != null) {
+                        failedTenant.closeAllConnections();
+                    }
+                });
+    }
+
+    @Override
+    public void addEndpoint(final String tenantId, final MqttEndpoint mqttEndpoint) {
+        getTenantConnections(tenantId).addEndpoint(mqttEndpoint);
+    }
+
+    @Override
+    public boolean closeEndpoint(final String tenantId, final MqttEndpoint mqttEndpoint) {
+
+        final boolean amqpLinkClosed = getTenantConnections(tenantId).closeEndpoint(mqttEndpoint);
+        if (amqpLinkClosed) {
+            connectionsPerTenant.remove(tenantId);
+        }
+
+        return amqpLinkClosed;
+    }
+
+    @Override
+    public void closeAllTenants() {
+        connectionsPerTenant.forEach((k, connections) -> connections.closeAllConnections());
+        connectionsPerTenant.clear();
+    }
+
+    @Override
+    public Future<TelemetrySender> getOrCreateTelemetrySender(final String tenantId) {
+        try {
+            return getAmqpAdapterClientFactory(tenantId).getOrCreateTelemetrySender();
+        } catch (Exception ex) {
+            return Future.failedFuture(ex);
+        }
+    }
+
+    @Override
+    public Future<EventSender> getOrCreateEventSender(final String tenantId) {
+        try {
+            return getAmqpAdapterClientFactory(tenantId).getOrCreateEventSender();
+        } catch (Exception ex) {
+            return Future.failedFuture(ex);
+        }
+    }
+
+    @Override
+    public Future<CommandResponder> getOrCreateCommandResponseSender(final String tenantId) {
+        try {
+            return getAmqpAdapterClientFactory(tenantId).getOrCreateCommandResponseSender();
+        } catch (Exception ex) {
+            return Future.failedFuture(ex);
+        }
+    }
+
+    @Override
+    public Future<MessageConsumer> createDeviceSpecificCommandConsumer(final String tenantId, final String deviceId,
+            final Consumer<Message> messageHandler) {
+
+        try {
+            return getAmqpAdapterClientFactory(tenantId).createDeviceSpecificCommandConsumer(deviceId, messageHandler);
+        } catch (Exception ex) {
+            return Future.failedFuture(ex);
+        }
+    }
+
+    @Override
+    public Future<MessageConsumer> createCommandConsumer(final String tenantId,
+            final Consumer<Message> messageHandler) {
+
+        try {
+            return getAmqpAdapterClientFactory(tenantId).createCommandConsumer(messageHandler);
+        } catch (Exception ex) {
+            return Future.failedFuture(ex);
+        }
+    }
+
+    private TenantConnections getTenantConnections(final String tenantId) throws IllegalArgumentException {
+        final TenantConnections tenantConnections = connectionsPerTenant.get(tenantId);
+        if (tenantConnections == null) {
+            throw new IllegalArgumentException("tenant [" + tenantId + "] is not connected");
```

Comment:
Throwing an exception here doesn't look that appropriate to me because it requires exception handling in all calling methods and would also require mention in the `MultiTenantConnectionManager` interface javadoc. 
How about returning a `Future<TenantConnections>` here and letting it fail if the map entry is `null`.
The `MultiTenantConnectionManager#addEndpoint` and `MultiTenantConnectionManager#closeEndpoint` could then also return a Future, and the "not connected/no connection registered" case could be handled via a failed Future.